### PR TITLE
Feature/market

### PR DESCRIPTION
**Pull request by friedger**
This PR
* adds a new account monstereosmt that manages the market
* adds new methods in monstereosio that allows to transfer monsters, and ensures that ram is paid by new owner
* simplifies the generation of abi/wast for the pet contract using petcode.c and petabi.c which include pet.c
* adds init script for market as well as script to test the market

### CODE REVIEW

File: services/eos-dev/contracts/pet/lib/pet.admin.cpp [hunk 1]
```diff
@@ -88,23 +88,20 @@ void pet::changepettyp(uint64_t id, vector<uint8_t> elements) {
...
     pc.last_element_id++;
-    eosio_assert(pc.last_element_id > 0, "_next_element_id overflow detected");
     _update_pet_config(pc);
     return pc.last_element_id-1; // zero based id
 }
 
 uint64_t pet::_next_pet_type_id(){
     auto pc = _get_pet_config();
     pc.last_pet_type_id++;
-    eosio_assert(pc.last_pet_type_id > 0, "_next_pet_type_id overflow detected");
```
Comment (leordev) [inline services/eos-dev/contracts/pet/lib/pet.admin.cpp]: please add this statement, it avoids overflow

File: services/eos-dev/contracts/pet/lib/pet.admin.cpp [hunk 2]
```diff
@@ -88,23 +88,20 @@ void pet::changepettyp(uint64_t id, vector<uint8_t> elements) {
...
     pc.last_id++;
-    eosio_assert(pc.last_id > 0, "_next_id overflow detected");
     _update_pet_config(pc);
     return pc.last_id;
 }
 
 uint64_t pet::_next_element_id(){
     auto pc = _get_pet_config();
     pc.last_element_id++;
-    eosio_assert(pc.last_element_id > 0, "_next_element_id overflow detected");
```
Comment (leordev) [inline services/eos-dev/contracts/pet/lib/pet.admin.cpp]: please add this statement, it avoids overflow

File: services/eos-dev/contracts/pet/lib/pet.admin.cpp [hunk 3]
```diff
@@ -88,23 +88,20 @@ void pet::changepettyp(uint64_t id, vector<uint8_t> elements) {
 uuid pet::_next_id(){
     auto pc = _get_pet_config();
     pc.last_id++;
-    eosio_assert(pc.last_id > 0, "_next_id overflow detected");
```
Comment (leordev) [inline services/eos-dev/contracts/pet/lib/pet.admin.cpp]: please add this statement, it avoids overflow

File: services/eos-dev/contracts/pet/lib/utils.hpp [hunk 4]
```diff
@@ -21,10 +21,15 @@ namespace utils {
         return p64[0] == 0 && p64[1] == 0 && p64[2] == 0 && p64[3] == 0;
     }
 
+    uint64_t hash_str(const string &str) {
```
Comment (leordev) [inline services/eos-dev/contracts/pet/lib/utils.hpp]: please remove function, we dont use it anymore

File: services/eos-dev/contracts/pet/lib/utils.hpp [hunk 5]
```diff
@@ -21,10 +21,15 @@ namespace utils {
...
+        return hash<string>{}(str);
+    }
+
     checksum256 get_hash(const bytes& data) {
         checksum256 result;
-        sha256(data.data(), data.size(), &result);
+        sha256((char *) data.data(), data.size(), &result);
```
Comment (leordev) [inline services/eos-dev/contracts/pet/lib/utils.hpp]: why did you add `(char *)` ?

File: services/eos-dev/contracts/pet/pet.cpp [hunk 6]
```diff
@@ -261,60 +340,6 @@ void pet::_update(st_pets &pet) {
...
 }
 
-// we need to sacrifice abi generation for recipient listener
-// keep alternating the comments between EOSIO_ABI (to generate ABI)
-// and EOSIO_ABI_EX to generate the listener action
-// https://eosio.stackexchange.com/q/421/54
-
-// EOSIO_ABI(pet, (createpet)(updatepet)(feedpet)(bedpet)(awakepet)(destroypet)(battlecreate)(battlejoin)(battleleave)(battlestart)(battleselpet)(battleattack)(battlefinish)(addelemttype)(changeelemtt)(addpettype)(changepettyp)(changecrtol)(changebatma)(changebatidt)(changebatami)(changebatama)(transfer))
-
-#define EOSIO_ABI_EX( TYPE, MEMBERS ) \
```
Comment (leordev) [inline services/eos-dev/contracts/pet/pet.cpp]: loved your idea of removing the comments and splitting in two cpp files, one for regular compilation and other one for abi, nice job! 👍

File: services/eos-dev/contracts/pet/lib/utils.hpp (hunk 5 above)
Comment (friedger) [reply services/eos-dev/contracts/pet/lib/utils.hpp]: The compiler was complaining..
